doc/adr/0001-jetstream-json-api-design.md

```diff
@@ -0,0 +1,131 @@
+# 1. JetStream JSON API Design
+
+Date: 2020-04-30
+
+## Status
+
+Proposed
+
+## Context
+
+At present, the API encoding consists of mixed text and JSON, we should improve consistency and error handling.
+
+## Decision
+
+We anticipate many types of consumer for JetStream including very small devices via MQTT and those written in 
+languages without easy to use JSON support like C.
+
+We will make it easy for light-weight clients by using `+OK` and `-ERR` style prefixes for client interactions,
+but the server management API will be JSON all the time. The rationale being that such small devices will most
+likely have their Consumers and Streams provisioned as part of the platform and often might not care for the
+additional context.
+
+In client responses we will use URL Encoding to provide additional context so multiple fields worth of
+context can be provided. URL Encoding has [broad language support](https://www.rosettacode.org/wiki/URL_decoding).
+
+### Admin APIs
+
+#### Requests
+
+All Admin APIs that today accept `nil` body should also accept an empty JSON document as request body.
+
+#### Responses
+
+All responses will be JSON objects, a few examples will describe it best. Any error that happens has to be 
+communicated within the originally expected message type. Even the case where JetStream is not enabled for
+an account, the response has to be the data type the API would normally generate with additional `error`.
+
+Successful Stream Info:
+
+```json
+{
+  "type": "io.nats.jetstream.api.v1.stream_info",
+  "time": "2020-04-23T16:51:18.516363Z",
+  "config": {
+    "name": "STREAM",
+    "subjects": [
+      "js.in"
+    ],
+    "retention": "limits",
+    "max_consumers": -1,
+    "max_msgs": -1,
+    "max_bytes": -1,
+    "max_age": 31536000,
+    "max_msg_size": -1,
+    "storage": "file",
+    "num_replicas": 1
+  },
+  "state": {
+    "messages": 95563,
+    "bytes": 40104315,
+    "first_seq": 34,
+    "last_seq": 95596,
+    "consumer_count": 1
+  }
+}
+```
+
+Stream Info Error:
+
+```json
+{
+  "type": "io.nats.jetstream.api.v1.stream_info",
+  "time": "2020-04-23T16:51:18.516363Z",
+  "error": {
+    "message": "unknown stream bob",
+    "code": 123,
+    "kind": "JetStreamBadRequest"
```

Comment:
We didn't discuss this field on the call - I see most have something similar, I dont really care for it so can remove unless someone else sees value.

Reply:
I do not think we need kind. We can always add later if we are wrong, harder to take it out.

Reply:
Is it common to use `message` vs `description`?

Reply:
I don't think there's any consensus about this stuff, whichever you prefer

Reply:
I think `description`, and `code` - since we overload message for other things.

Reply:
The `kind` thing is to let us indicate to things like a web ui what the error is - it cant parse the body of the description. And looks like we'll use `code` in the HTTP sense to indicate user/server error.

So still not sure if we need it, but that was my angle with it.